stores.js

```diff
@@ -0,0 +1,301 @@
+          } else {
+            found = true
+            cb(null, result.value)
+            queryStream.destroy()
+          }
+        }
+      }
+    })
+    if (value && streamMode) {
+      // push local cached value to stream first
+      responseStream.push(value)
+    }
+    pipeline(queryStream, responseStream, (err) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      if (streamMode === false && found === false) {
+        cb(null, null)
+      }
+    })
+    if (streamMode) return responseStream
+  }
+  put (value, cb) {
+    assert(typeof cb === 'function', 'Callback is required')
+    assert(Buffer.isBuffer(value), 'Value must be a buffer')
+    assert(
+      value.length <= PUT_VALUE_MAX_SIZE,
+      `Value size must be <= ${PUT_VALUE_MAX_SIZE}`
+    )
+    const { store, dht } = this
+    const key = Buffer.alloc(32)
+    hash(key, value)
+    // set locally for easy cached retrieval
+    store.set(key.toString('hex'), value)
+
+    // send to the dht
+    const queryStream = dht.update('immutable-store', key, value)
+
+    queryStream
+      .once('warning', (err = {}) => {
+        if (err.message === 'ERR_INVALID_INPUT') {
+          queryStream.destroy(err)
+        }
+      })
+
+    queryStream.resume()
+
+    finished(queryStream, (err) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      cb(null, key)
+    })
+  }
+  _command () {
+    const { store } = this
+    return {
+      update ({ target, value }, cb) {
+        const key = Buffer.alloc(32)
+        hash(key, value)
+        if (Buffer.compare(key, target) !== 0) {
+          cb(Error('ERR_INVALID_INPUT'))
+          return
+        }
+        store.set(key.toString('hex'), value)
+        cb(null)
+      },
+      query ({ target }, cb) {
+        cb(null, store.get(target.toString('hex')))
+      }
+    }
+  }
+}
+class MutableStore {
+  constructor (dht, store) {
+    this.dht = dht
+    this.store = store
+  }
+  salt (size = 32) {
+    assert(
+      size >= 16 && size <= 64,
+      'salt size must be between 16 and 64 bytes (inclusive)'
+    )
+    const salt = Buffer.alloc(size)
+    randomBytes(salt)
+    return salt
+  }
+  keypair () {
+    const publicKey = Buffer.alloc(pkSize)
+    const secretKey = Buffer.alloc(skSize)
+    createKeypair(publicKey, secretKey)
+    return { publicKey, secretKey }
+  }
+  get (key, opts = {}, cb = opts) {
+    const { dht } = this
+    const { salt, seq = 0 } = opts
+    assert(typeof seq === 'number', 'seq should be a number')
+    if (salt) {
+      assert(Buffer.isBuffer(salt), 'salt must be a buffer')
+      assert(
+        salt.length >= 16 && salt.length <= 64,
+        'salt size must be between 16 and 64 bytes (inclusive)'
+      )
+    }
+    const queryStream = dht.query('mutable-store', key, { salt, seq })
+    let found = false
+    const userSeq = seq
+    const streamMode = typeof cb !== 'function'
+    const responseStream = Transform({
```

Comment:
I just remembered we have a map option in dht-rpc for exactly this purpose (ie value transform)!

We should use that instead of the transform, query(..., { map(data) { return newData } })

Reply:
See here for an example, https://github.com/hyperswarm/dht/blob/master/index.js#L48

Getting the dht node out as well is super helpful like we do in the peer lookup :)